File: CPlusPlus/OOPS/Polymorphism.cpp
/*The word polymorphism means having many forms.
* In simple words, we can define polymorphism as the ability of a message to be displayed in more than one form.
*/


/* This program also gives an example about function overriding
*  Function overriding occurs when a derived class has a definition for one of the member functions of the base class. That base function is said to be overridden.
*/
   
  
#include <bits/stdc++.h>
using namespace std;
  
class base
{
public:
    virtual void print ()
    { cout<< "print base class" <<endl; }
   
    void show ()
    { cout<< "show base class" <<endl; }
};
   
class derived:public base
{
public:
    void print () 
//print () is already virtual function in derived class, we could also declared as virtual void print () explicitly
    { cout<< "print derived class" <<endl; }
   
    void show ()
    { cout<< "show derived class" <<endl; }
};
  
//main function
int main() 
{
    base *bptr;
    derived d;
    bptr = &d;
       
    //virtual function, binded at runtime (Runtime polymorphism)
    bptr->print(); 
       
    // Non-virtual function, binded at compile time
    bptr->show(); 
  
    return 0;
} 


/*This program gives the output as:-
print derived class
show base class
*/

File: Java/OOPS/Polymorphism.java
/*The word polymorphism means having many forms. In simple words, we can define polymorphism as the ability of a message to be displayed in more than one form.*/

class Polymorphism{  
    public static void main(String args[]){  
        Bank b;  
        b=new SBI();  // Created an object of SBI
        System.out.println("SBI Rate of Interest: "+b.getRateOfInterest());  // The getRateOfInterest function of SBI class is called
        b=new ICICI();  
        System.out.println("ICICI Rate of Interest: "+b.getRateOfInterest());  // The getRateOfInterest function of ICICI class is called
        b=new AXIS();  
        System.out.println("AXIS Rate of Interest: "+b.getRateOfInterest());  // The getRateOfInterest function of AXIS class is called
    }  
}

class Bank{ 
    float getRateOfInterest(){return 0;}  
} 
 
class SBI extends Bank{  
    float getRateOfInterest(){return 8.4f;}  
}  

class ICICI extends Bank{  
    float getRateOfInterest(){return 7.3f;}  
}  

class AXIS extends Bank{  
    float getRateOfInterest(){return 9.7f;}  
}    

/*
    OUTPUT:
        SBI Rate of Interest: 8.4
        ICICI Rate of Interest: 7.3
        AXIS Rate of Interest: 9.7

*/
